ambari-server/src/main/java/org/apache/ambari/server/orm/entities/BlueprintMpackReferenceEntity.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ambari.server.orm.entities;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import javax.persistence.CascadeType;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
+import javax.persistence.Table;
+import javax.persistence.TableGenerator;
+
+@Entity
+@Table(name = "blueprint_mpack_reference")
+@TableGenerator(name = "blueprint_mpack_reference_id_generator", table = "ambari_sequences", pkColumnName = "sequence_name",
+  valueColumnName = "sequence_value", pkColumnValue = "blueprint_mpack_ref_id_seq", initialValue = 1)
+public class BlueprintMpackReferenceEntity {
+  @Id
+  @GeneratedValue(strategy = GenerationType.TABLE, generator = "blueprint_mpack_reference_id_generator")
+  @Column(name = "id", nullable = false, updatable = false)
+  private Long id;
+
+  @Column(name = "mpack_name")
```

Comment:
Why do we have duplicated fields here? Shouldn't this just be a foreign key to an mpack?

Reply:
I wanted to decouple the blueprint lifecycle from the mpack lifecycle. This is a valid case:

- user posts a blueprint (mpack does not exist at this time as automatic mpack download will happen at cluster installation time)
- user installs the necessary mpacks (or they are auto-dowloaded at cluster installation)

I am not sure this can be modelled with foreign keys as after the first step there are no mpacks in the DB the FK's can refer to.

Reply:
Is that really a valid use case? Ambari can't auto download the correct management pack because there could be multiple places it's hosted. The mpacks have to be installed first for any blueprint installation, right?

Reply:
I think so, but let's ask Jayush.

Reply:
@rnettleton @jayush  Could you pls. clarify if blueprints have an independent lifecycle from mpacks, such as the user can send in a blueprint referencing an mpack that is not yet installed.

Reply:
@benyoka , Blueprints should indeed have a separate lifecycle from mpack installs.  

A POST of a Blueprint to the Ambari REST API should only cause the document to the validated and persisted.  Since the Blueprint can include references to an MPack, these should be validated as much as possible, but no download/install of an MPack should occur due to just a POST of a Blueprint.  The MPack installation process, when handled by Blueprints, should only occur when a Cluster Creation Template (referring to a Blueprint) is POST-ed to the Ambari Server.  

@jonathan-hurley , We're trying to support multiple deployment models in Blueprints: 

1. Ambari user manually installs a set of mpacks, and then deploys a Blueprint that references these. 
2. Ambari user lists the necessary MPacks in a Blueprint (including download URLs), and expects Ambari to handle the downloading/installation of the mpacks prior to the cluster creation start.

Reply:
So in the last case, where the mpacks are not already installed, you are saying that you will only pass some basic info about an mpack (which can't be trusted since the only source of truth would be the metadata inside of the downloaded mpack bundle) along with a URL (the only source of truth). In this scenario, nothing is downloaded yet, right? 

So, the FK can be empty, but once the mpack is downloaded, should it not have an association made? At the very least, we shouldn't put names in here since they could be wrong compared to what's actually in the mpack.

Reply:
In the last case, when a blueprint is saved the only things it knows about an mpack is its name and version. I need to store this information hence this table with its mpack_name/mpack_version fields.

I also need to store/link information that is not directly related to the mpack but to the blueprint's use of that mpack, e.g. special configuration that applies to services installed from that mpack.

Would you agree with the following changes?
- Adding a nullable FK column that references the mpack table (still need to store mpack_name and mpack_version as those will be the basis of finding the right mpack)
- renaming the entity to MpackInstanceEntity as Mpack**Reference**Entity can be misleading.

Reply:
Yes, I'd be good with those changes - and to also document that the mpack name/version is what was supplied and may not be correct (or even change their field names to something that reflects that fact)

Reply:
+1 for adding a nullable `mpack_id`